Use local files create database

marca116 opened the pull request:
Allows the use of local pdf files for the creation of the database, including using the path to a folder containing pdf files.
Saves the files in a static folder, in order for them to be accessible from within the browser. Otherwise, the browser blocks access to local files on the computer.

Issue: #15

pierreolivierbonin commented:
Is this the relevant doc for what you've built here? https://docs.streamlit.io/develop/concepts/configuration/serving-static-files

pierreolivierbonin commented:
Could you please further document how to set things up for the local pdfs to be working? Some explaining in the README would be useful.

pierreolivierbonin commented:
I'm thinking perhaps we should have a collection created and named 'local_collection' or something, which would be based entirely on the pdfs found in `static/` and ingested into the vector database. What do you think?

marca116 commented:
> Could you please further document how to set things up for the local pdfs to be working? Some explaining in the README would be useful.
> 
> For instance, we could explain with a quick example what to replace in `db_config.py` to make it work - or whether we need to change anything at all? If just putting the files in the `static` folder in root is sufficient, then we should make it clear.

It's not necessary to put anything in the static folder manually. The extract_pdf script should automatically copy/download any file included in the db_config database setup.

I could potentially add a section in the readme to explain how to create your own database by modifying the db config. I could also clarify that any pdf files included will be available to the users with an app/static url.

marca116 commented:
> I'm thinking perhaps we should have a collection created and named 'local_collection' or something, which would be based entirely on the pdfs found in `static/` and ingested into the vector database. We could set it up so it gets automatically created if files are present in the folder, and if not, then it does not get created.
> 
> What do you think?

I'm not quite sure what would be the point, given that the static folder is populated automatically by using the script. The users shouldn't need to modify that folder, unless they want to delete a database and remove all files related to it.

marca116 commented:
> I also found these lines in the console, which are new to me, and some research directed me to an issue where other folks are encountering the same issue:
> 
> ```shell
> Failed to send telemetry event ClientStartEvent: capture() takes 1 positional argument but 3 were given
> Failed to send telemetry event ClientCreateCollectionEvent: capture() takes 1 positional argument but 3 were given
> ```
> 
> Even though it may be out of scope for this PR, I think we should probably investigate, as per [vanna-ai/vanna#917](https://github.com/vanna-ai/vanna/issues/917)

You can create an issue for it, and I could take a look at it.

marca116 commented:
I've modified the readme to include clear instructions on how to create your own database.

pierreolivierbonin commented on src/context.py:
```diff
@@ -2,7 +2,7 @@
 from typing import List
 import streamlit as st
-from transformers import PreTrainedTokenizerFast
+from transformers.tokenization_utils_fast import PreTrainedTokenizerFast
```
Does this fix a warning or something?

marca116 replied:
Yes, it did to me. I think the old import was deprecated.

pierreolivierbonin commented:
As requested, the README instructs that the system supports folder paths to include all PDFs in a directory. However, when I take a look at `db_config.py`, it takes me some time to understand how to write the config for a custom use case of local pdfs within the `VectorDBDataFiles` dataclass. 

As a first-time user of this feature - and without actually testing it, in order to test whether it's an intuitive approach or not - I would be inclined to writing something like the following in the dataclass, assuming that the "en" key of the dictionary within the "pdf" key of this custom database dictionary accepts a local folder path as well as the paths of individual files:

```sh
@dataclass
class VectorDBDataFiles:
    databases = [
...
        {
            "name": "equity",
            "languages": ["en", "fr"],
            "pdf": {
...
        {
            "name": "another_use_case",
            "languages": ["en"],
            "pdf": {
                "en": ["../use_case_x/"]  #<----- This would be the path of the local folder where I would include all my pdfs
            }
        }
```

Can you confirm whether this is correct? And if not, what I should do instead?

I have a feeling that this is getting a little messy and we could probably make a better use of the `VectorDBDataFiles` dataclass, using more attributes, etc.

marca116 commented:
> So after actually testing it and changing the dictionary keys and values in `db_config`, I'm getting the following traceback:
> 
> ```shell
> ────────────────────────── Traceback (most recent call last) ───────────────────────────
>   D:\Work\Canada-Labour-Research-Assistant\.venv\Lib\site-packages\streamlit\runtime\s
>   criptrunner\exec_code.py:121 in exec_func_with_error_handling
> 
> 
> 
>   D:\Work\Canada-Labour-Research-Assistant\.venv\Lib\site-packages\streamlit\runtime\s
>   criptrunner\script_runner.py:645 in code_to_exec
> 
> 
> 
>   D:\Work\Canada-Labour-Research-Assistant\chatbot_app.py:376 in <module>
> 
> 
> 
>     373 │
> 
>     374 │   else:
> 
>     375 │   │   app = App(is_remote=is_remote, engine=engine, hyperparams=hyperp
>   D:\Work\Canada-Labour-Research-Assistant\chatbot_app.py:96 in sidebar_config
> 
>      93 │   │   │
>      94 │   │   │   selected_db_name = st.selectbox(label=self.translator.get('sidebar
>      95 │   │   │   │   │   │   │   │   options=list(translated_db_list),
>   ❱  96 │   │   │   │   │   │   │   │   index=translated_db_list.index(translated_db_n
>      97 │   │   │
>      98 │   │   │   self.db_name = orig_db_list[translated_db_list.index(selected_db_n
>      99
> ────────────────────────────────────────────────────────────────────────────────────────
> ValueError: 'Labour' is not in list
> ```
> 
> Investigating leads me back to `resources/en.json` and the `fr.json` version. I then have to change the keys and values on lines 36+, then it still does not work.
> 
> I then also have to change line 53 in `config.py` file to finally make it work.
> 
> Altogether, I have to change values in **three different config files**, which is overall a frustrating developer experience. We need to streamline this a lot more so the user can make this work by changing not 3 config files, but only one file. And ideally, making it clear where the list within a dictionary within another dictionary can accept both the strings of individual files or just the path to a local folder.

I agree that this process is not very intuitive. For one, I think the translated database name should be set in db config, in the database config. I'll add a new key for that, and add instructions in the Readme on how to do that.

For the default database, I could add a field 'is_default' on the database, and read that when creating the interface. I think that would solve the issue. I'll also add an example with all 3 possibilities in one in the config, this should help clarify things.

marca116 commented:
I've finished making the previously mentioned modification. It should now be possible to modify the databases list without editing the resources or the chatbot interface config file.

I will look into the issue with the double asterix, but I don't think it's relevant to this PR.

marca116 commented:
> We're getting closer to completion and this is looking pretty good! One thing I just noticed is that some hyperlinks presented in the "metadata" tab are not correctly displayed for interactivity. For example:
> 
> ![image](https://private-user-images.githubusercontent.com/31133836/462572559-de49fd10-8749-4e7e-9c44-d0345e1cd6d9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE2NjEyOTAsIm5iZiI6MTc1MTY2MDk5MCwicGF0aCI6Ii8zMTEzMzgzNi80NjI1NzI1NTktZGU0OWZkMTAtODc0OS00ZTdlLTljNDQtZDAzNDVlMWNkNmQ5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzA0VDIwMjk1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlkZmYyYjVhM2U2ZjAwM2M2YmEwODE5NDQ4ZDU0YjY0MzdlMGI2MWM3MGVhYzE5OWQ5ZDc0MzU5MmJiMGUwMTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.IbPv4ptHNOinTabiyJckIfrNABPd-xSYTqeoSOaRWoE)
> 
> I have checked and it's not the file that is broken; it works when I manually add '#page=5' and reload the page in the browser, as an example.

I think the issue here is that the link is not recognized due to the whitespaces. I think replacing whitespaces with %20 in the underlying url should fix it.

pierreolivierbonin commented:
There's issues with the formatting of direct quotations. 

MRE: question asked on Ollama with gemma3n:latest --> "how can i use my employment equity data?"

Behaviour observed: first few quotations are displayed correctly, then I get the broken formatting below...

Suggestion: let's just remove all bullets, italics, bold formatting etc. Some regex that would take care of removing all punctuation except for period (".") may be a solution. Also something that would lowercase all characters, and then capitalize only the first one following a period could help clean these out and make it more consistent across the board.

![image](https://github.com/user-attachments/assets/ed3850ee-3807-4bfd-b3da-1f518248e6f3)

marca116 commented:
> There's issues with the formatting of direct quotations.
> 
> MRE: question asked on Ollama with gemma3n:latest --> "how can i use my employment equity data?"
> 
> Behaviour observed: first few quotations are displayed correctly, then I get the broken formatting below...
> 
> Suggestion: let's just remove all bullets, italics, bold formatting etc. Some regex that would take care of removing all punctuation except for period (".") may be a solution. Also something that would lowercase all characters, and then capitalize only the first one following a period could help clean these out and make it more consistent across the board.
> 
> ![image](https://private-user-images.githubusercontent.com/31133836/462833810-ed3850ee-3807-4bfd-b3da-1f518248e6f3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE5OTY2NzYsIm5iZiI6MTc1MTk5NjM3NiwicGF0aCI6Ii8zMTEzMzgzNi80NjI4MzM4MTAtZWQzODUwZWUtMzgwNy00YmZkLWIzZGEtMWY1MTgyNDhlNmYzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzA4VDE3MzkzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA5ZGRjZTQxNWQxZDgyMmY5ZmFiZjhlM2Q2Yzk1Mzk3NTk3MzRlZjM2NjI4NGNhMWEyMTJhYmM4OWUyMWZmNDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ocf1Y8_aVFwfds7dApSOMfmdADiSRMqwmTgT1Epcdko)

The reason for the capitalized letters is because that's how I differentiated the headers from the rest of the text. However, I see that it doesn't always look the best, and adds confusion about why some parts are capitalized and others not. 

To fix this, I changed the headers to instead use markdown emphasis (**...**). I also cleaned various other kinds of markdown that occurred for the pdf files, which made the formatting inconsistent. Finally, since I suspect that emphasis was better represented as capitalized letters in the embeddings, I replaced any text within ** with capitalized letters before embedding them. This means that the results of queries to the vector database should remain the same as they were before. Overall, I do feel like the visual formatting is a lot better than it was previously. Keep in mind that you will need to rerun the database creation script for the formatting to be updated.
